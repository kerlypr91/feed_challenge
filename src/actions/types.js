// Post List
export const GET_POST_LIST = 'GET_POST_LIST'
export const GET_POST_LIST_IN_PROGRESS = 'GET_POST_LIST_IN_PROGRESS'
export const GET_POST_LIST_SUCCESS = 'GET_POST_LIST_SUCCESS'
export const GET_POST_LIST_FAILURE = 'GET_POST_LIST_FAILURE'

// Comments Per Post
export const GET_POST_COMMENTS = 'GET_POST_COMMENTS'
export const GET_POST_COMMENTS_IN_PROGRESS = 'GET_POST_COMMENTS_IN_PROGRESS'
export const GET_POST_COMMENTS_SUCCESS = 'GET_POST_COMMENTS_SUCCESS'
export const GET_POST_COMMENTS_FAILURE = 'GET_POST_COMMENTS_FAILURE'

// Add New Comment
export const ADD_POST_COMMENT = 'ADD_POST_COMMENT'
export const ADD_POST_COMMENT_IN_PROGRESS = 'ADD_POST_COMMENT_IN_PROGRESS'
export const ADD_POST_COMMENT_SUCCESS = 'ADD_POST_COMMENT_SUCCESS'
export const ADD_POST_COMMENT_FAILURE = 'ADD_POST_COMMENT_FAILURE'