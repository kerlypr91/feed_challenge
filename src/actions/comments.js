import {
  GET_POST_COMMENTS,
  ADD_POST_COMMENT
} from './types'

// Get Post List Action - dispatches saga
export function getPostCommentsAction (postId) {
  return {
    type: GET_POST_COMMENTS,
    payload: { postId }
  }
}

export function addPostCommentAction (postId, comment) {
  return {
    type: ADD_POST_COMMENT,
    payload: { postId, comment }
  }
}

