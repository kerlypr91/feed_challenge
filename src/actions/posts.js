import { GET_POST_LIST } from './types'

// Get Post List Action - dispatches saga
export function getPostListAction () {
  return {
    type: GET_POST_LIST,
    payload: {}
  }
}
