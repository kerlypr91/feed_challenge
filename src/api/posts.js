import { createApi } from './index'

export function getPostListAPI () {
  return createApi().get('https://jsonplaceholder.typicode.com/posts')
}