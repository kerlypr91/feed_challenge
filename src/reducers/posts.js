import {
  GET_POST_LIST_IN_PROGRESS,
  GET_POST_LIST_SUCCESS,
  GET_POST_LIST_FAILURE
} from '../actions/types'

export const initialState = {
  postList: [],
  fetching: false,
  error: null
}

export default function (state = initialState, action) {
  switch (action.type) {
    case GET_POST_LIST_IN_PROGRESS:
      return {
        ...state,
        fetching: true,
        error: null
      }
    case GET_POST_LIST_SUCCESS:
      const { payload } = action

      return {
        ...state,
        postList: payload,
        fetching: false,
        error: null
      }
    case GET_POST_LIST_FAILURE:
      return {
        ...state,
        fetching: false,
        error: { ...action.payload }
      }
    default:
      return state
  }
}
