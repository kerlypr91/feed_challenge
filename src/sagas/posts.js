import { put, call, delay } from 'redux-saga/effects'
import {
  GET_POST_LIST_IN_PROGRESS,
  GET_POST_LIST_SUCCESS,
  GET_POST_LIST_FAILURE
} from '../actions/types'
import { getPostListAPI } from '../api/posts'

export function* fetchPostListSaga (action) {
  yield put({ type: GET_POST_LIST_IN_PROGRESS })

  yield delay(800)

  try {
    const response = yield call(getPostListAPI)

    const { data, ok } = response

    if (ok) {
      yield put({ type: GET_POST_LIST_SUCCESS, payload: data })
    }
  } catch (e) {
    yield put({ type: GET_POST_LIST_FAILURE, payload: e })
  }
}
