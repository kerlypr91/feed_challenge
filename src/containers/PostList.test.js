import React from 'react';
import { shallow } from 'enzyme';
import {
  PostList,
  mapStateToProps,
  mapDispatchToProps
} from './PostList';
import * as PostsActions from '../actions/posts'

describe('PostList Tests', () => {
  let wrapper
  let getPostList
  let postList
  let history
  let setStateSpy

  beforeEach(() => {
    postList = [
      {
        id: 1,
        title: 'Title 1',
        body: 'Body 1'
      },
      {
        id: 2,
        title: 'Title 2',
        body: 'Body 2'
      }
    ]

    history = {
      push: jest.fn()
    }

    getPostList = jest.fn()

    setStateSpy = jest.spyOn(PostList.prototype, 'setState')

    wrapper = shallow(<PostList getPostList={getPostList} history={history} />)
  })

  it('should match snapshot - loader container', () => {
    const loaderContainer = wrapper.find('.loader_container')

    expect.assertions(2)
    expect(loaderContainer).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot - post list container', () => {
    wrapper.setProps({
      postList
    })

    wrapper.setState({
      loading: false
    })

    wrapper.update()

    const postListContainer = wrapper.find('.post_list')

    expect.assertions(2)
    expect(postListContainer).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('componentDidMount calls prop getPostList', () => {
    wrapper.instance().componentDidMount()

    expect(getPostList).toHaveBeenCalled()
  })

  it('componentDidUpdate calls setState and sets loading value to false', () => {
    wrapper.setProps({
      fetching: false,
      error: null,
      postList
    })

    wrapper.instance().componentDidUpdate({ fetching: true })

    expect.assertions(2)
    expect(setStateSpy).toHaveBeenCalled()
    expect(wrapper.state('loading')).toBe(false)
  })

  it('handleCommentLinkClick should navigate to comments route', () => {
    wrapper.instance().handleCommentLinkClick(1)()

    expect.assertions(2)
    expect(history.push).toHaveBeenCalled()
    expect(history.push).toHaveBeenCalledWith('/comments', { postId: 1 })
  })

  it('mapDispatchToProps: getPostList action', () => {
    const dispatch = jest.fn()
    const spyOn = jest.spyOn(PostsActions, 'getPostListAction')

    mapDispatchToProps(dispatch).getPostList()

    expect(spyOn).toHaveBeenCalled()
  })

  it('mapStateToProps', () => {
    const state = {
      posts: {
        postList: 'test',
        fetching: true,
        error: 'error'
      }
    }

    expect.assertions(7)
    expect(mapStateToProps(state)).toBeTruthy()
    expect(mapStateToProps(state)).toHaveProperty('postList')
    expect(mapStateToProps(state).postList).toBe('test')
    expect(mapStateToProps(state)).toHaveProperty('fetching')
    expect(mapStateToProps(state).fetching).toBe(true)
    expect(mapStateToProps(state)).toHaveProperty('error')
    expect(mapStateToProps(state).error).toBe('error')
  })
})