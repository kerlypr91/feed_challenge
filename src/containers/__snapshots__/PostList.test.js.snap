// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList Tests should match snapshot - loader container 1`] = `
<div
  className="loader_container"
>
  <Loader />
</div>
`;

exports[`PostList Tests should match snapshot - post list container 1`] = `
<Fragment>
  <div
    className="post_title"
  >
    <i
      className="fa fa-reddit"
    />
        POSTS FEED
  </div>
  <div
    className="post_list"
  >
    <Post
      body="Body 1"
      id={1}
      key="1"
      onCommentLinkClick={[Function]}
      title="Title 1"
    />
    <Post
      body="Body 2"
      id={2}
      key="2"
      onCommentLinkClick={[Function]}
      title="Title 2"
    />
  </div>
</Fragment>
`;
